File: Correspondences.lua
local ffi = require 'ffi'
local torch = require 'torch'
local utils = require 'pcl.utils'
local pcl = require 'pcl.PointTypes'

local Correspondences = torch.class('pcl.Correspondences', pcl)

local methods = {}

local function init()
  local Correspondences_method_names = {
    'new',
    'clone',
    'delete',
    'size',
    'at',
    'push_back',
    'pop_back',
    'clear',
    'insert',
    'erase',
    'empty'
  }
  
  -- use utility function to automatically register new/clone results for GC delete
  methods = utils.create_typed_methods("pcl_Correspondences_", Correspondences_method_names, '')
end

init()

function Correspondences:__init()
  rawset(self, 'f', methods)
  rawset(self, 'o', self.f.new())
end

function Correspondences:cdata()
  return self.o
end

function Correspondences:clone()
  return self.f.clone(self.o)
end

function Correspondences:size()
  return self.f.size(self.o)
end

function Correspondences:__len()
  return self:size()
end

function Correspondences:__index(idx)
  local v = rawget(self, idx)
  if not v then 
    v = Correspondences[idx]
    if not v and type(idx) == 'number' then
      local f, o = rawget(self, 'f'), rawget(self, 'o')
      v = f.at(o, idx-1)
    end
  end
  return v
end

function Correspondences:__newindex(idx, v)
  local f, o = rawget(self, 'f'), rawget(self, 'o')
  if type(idx) == 'number' then
    f.at(o, idx-1):set(v)
  else
    rawset(self, idx, v)
  end
end

function Correspondences:push_back(value)
  self.f.push_back(self.o, value)
end

function Correspondences:pop_back()
  return self.f.pop_back(self.o)
end

function Correspondences:clear()
  self.f.clear(self.o)
end

function Correspondences:insert(pos, value)
  self.f.insert(self.o, pos-1, value)
end

function Correspondences:erase(begin_pos, end_pos)
  self.f.erase(self.o, begin_pos-1, (end_pos or begin_pos + 1)-1)
end

function Correspondences:empty()
  return self.f.empty(self.o)
end

function Correspondences:__pairs()
  return function (t, k)
    local i = k or 1
    if i > #t then
      return nil
    else
      local v = t[i]
      return i+1, v
    end
  end, self, nil
end

function Correspondences:__ipairs()
  return self:__pairs()
end

function Correspondences:__tostring()
  local t = {}
  for i,v in ipairs(self) do
    table.insert(t, tostring(v))
  end
  table.insert(t, string.format("[pcl.Correspondences of size %d]", #self))
  return table.concat(t, '\n')
end

File: src/correspondences.cpp
#include "utils.h"
#include <pcl/correspondence.h>

#define TYPE_KEY _

struct _Correspondences
{
  int index_query;
  int index_match;

  union
  {
    float distance;
    float weight;
  };
};


PCLIMP(Correspondences_ptr *, Correspondences, new)()
{
  return new Correspondences_ptr(new pcl::Correspondences());
}

PCLIMP(Correspondences_ptr *, Correspondences, clone)(Correspondences_ptr *self)
{
  return new Correspondences_ptr(new pcl::Correspondences(**self));
}

PCLIMP(void, Correspondences, delete)(Correspondences_ptr *self)
{
  delete self;
}

PCLIMP(int, Correspondences, size)(Correspondences_ptr *self)
{
  return static_cast<int>((*self)->size());
}

PCLIMP(_Correspondences, Correspondences, getAt)(Correspondences_ptr *self, size_t pos)
{
  pcl::Correspondences &v = **self;
  const pcl::Correspondence &c = v[pos];
  _Correspondences c_ = { c.index_query, c.index_match, c.distance };
  return c_;
}

PCLIMP(void, Correspondences, setAt)(Correspondences_ptr *self, size_t pos, const _Correspondences &value)
{
  pcl::Correspondences &v = **self;
  v[pos] = pcl::Correspondence(value.index_query, value.index_match, value.distance);
}

PCLIMP(void, Correspondences, push_back)(Correspondences_ptr *self, const _Correspondences &value)
{
  (*self)->push_back(pcl::Correspondence(value.index_query, value.index_match, value.distance));
}

PCLIMP(void, Correspondences, pop_back)(Correspondences_ptr *self)
{
  (*self)->pop_back();
}

PCLIMP(void, Correspondences, clear)(Correspondences_ptr *self)
{
  (*self)->clear();
}

PCLIMP(void, Correspondences, insert)(Correspondences_ptr *self, size_t pos, const _Correspondences &value)
{
  pcl::Correspondences &v = **self;
  pcl::Correspondences::iterator i = pos >= v.size() ? v.end() : v.begin() + pos;
  v.insert(i, pcl::Correspondence(value.index_query, value.index_match, value.distance));
}

PCLIMP(void, Correspondences, erase)(Correspondences_ptr *self, size_t begin, size_t end)
{
  if (begin >= end)
    return;

  pcl::Correspondences& v = **self;
  pcl::Correspondences::iterator b = begin >= v.size() ? v.end() : v.begin() + begin;
  pcl::Correspondences::iterator e = end >= v.size() ? v.end() : v.begin() + end;
  v.erase(b, e);
}

PCLIMP(bool, Correspondences, empty)(Correspondences_ptr *self)
{
  return (*self)->empty();
}
